cmd/simulator/txs/agent.go

```diff
@@ -53,33 +59,80 @@ func (a issueNAgent[T]) Execute(ctx context.Context) error {
 	}
 
 	txChan := a.sequence.Chan()
+	confirmedCount := 0
+	batchI := 0
+	m := a.metrics
+
+	// Tracks the total amount of time waiting for issuing and confirming txs
+	var (
+		totalIssuedTime    time.Duration
+		totalConfirmedTime time.Duration
+	)
+
+	defer func() {
+		if err := a.worker.Close(ctx); err != nil {
+			log.Error("error trying to close worker: %w", "err", err)
+		}
+	}()
 
+	// Start time for execution
+	start := time.Now()
 	for {
 		var (
-			txs = make([]T, 0, a.n)
-			tx  T
-			ok  bool
+			txs     = make([]T, 0, a.n)
+			tx      T
+			moreTxs bool
 		)
+		// Start issuance batch
+		issuedStart := time.Now()
+	L:
 		for i := uint64(0); i < a.n; i++ {
 			select {
-			case tx, ok = <-txChan:
-				if !ok {
-					return a.worker.Close(ctx)
-				}
 			case <-ctx.Done():
 				return ctx.Err()
+			case tx, moreTxs = <-txChan:
+				if !moreTxs {
+					break L
+				}
+				issuanceIndividualStart := time.Now()
+				if err := a.worker.IssueTx(ctx, tx); err != nil {
+					return fmt.Errorf("failed to issue transaction %d: %w", len(txs), err)
+				}
+				issuanceIndividualTime := time.Since(issuanceIndividualStart)
+				m.IssuanceTxTimes.Observe(issuanceIndividualTime.Seconds())
+				txs = append(txs, tx)
 			}
-
-			if err := a.worker.IssueTx(ctx, tx); err != nil {
-				return fmt.Errorf("failed to issue transaction %d: %w", len(txs), err)
-			}
-			txs = append(txs, tx)
 		}
+		// Get the batch's issuance time and add it to totalIssuedTime
+		issuedDuration := time.Since(issuedStart)
+		log.Info("Issuance Batch Done", "batch", batchI, "time", issuedDuration.Seconds())
+		totalIssuedTime += issuedDuration
 
+		// Wait for txs in this batch to confirm
+		confirmedStart := time.Now()
 		for i, tx := range txs {
+			confirmedIndividualStart := time.Now()
 			if err := a.worker.ConfirmTx(ctx, tx); err != nil {
 				return fmt.Errorf("failed to await transaction %d: %w", i, err)
 			}
+			confirmationIndividualTime := time.Since(confirmedIndividualStart)
+			m.ConfirmationTxTimes.Observe(confirmationIndividualTime.Seconds())
```

Comment:
The confirmation time should be the time from issuing the transaction to when it is confirmed.

The current metric captures the amount of time it takes from starting to try and confirm the transaction (which is after the transaction has already been issued) to when we've finished confirming it.

Reply:
fixed